File: backend/app/models/trip.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime


class TripMember(BaseModel):
    userId: Optional[str] = None
    email: Optional[str] = None
    displayName: str
    photoURL: Optional[str] = None
    homeCurrency: str
    role: str  # "owner" | "member" | "ghost"
    joinedAt: Optional[datetime] = None
    ghostId: Optional[str] = None


class TripInvite(BaseModel):
    email: str
    invitedBy: str
    status: str = "pending"


class CustomCategory(BaseModel):
    id: str
    name: str
    emoji: str = "🏷️"
    createdBy: str


class TripCreate(BaseModel):
    name: str
    destination: str
    destinationCurrency: Optional[str] = None
    startDate: str
    endDate: str
    budget: Optional[float] = None
    budgetCurrency: Optional[str] = None


class TripUpdate(BaseModel):
    name: Optional[str] = None
    destination: Optional[str] = None
    startDate: Optional[str] = None
    endDate: Optional[str] = None
    budget: Optional[float] = None
    budgetCurrency: Optional[str] = None


class GhostMemberCreate(BaseModel):
    displayName: str
    homeCurrency: str


class GhostMemberUpdate(BaseModel):
    displayName: Optional[str] = None
    homeCurrency: Optional[str] = None


class GhostPromote(BaseModel):
    email: str


class InviteCreate(BaseModel):
    email: str


class CustomCategoryCreate(BaseModel):
    name: str
    emoji: str = "🏷️"

File: android/app/src/main/java/com/duitrip/app/data/model/Trip.kt
package com.duitrip.app.data.model

import com.google.firebase.Timestamp

data class TripMember(
    val userId: String? = null,
    val email: String? = null,
    val displayName: String = "",
    val photoURL: String? = null,
    val homeCurrency: String = "USD",
    val role: String = "member", // "owner" | "member" | "ghost"
    val joinedAt: Timestamp? = null,
    val ghostId: String? = null,
) {
    /** userId for real members, ghostId for ghosts — the id used across splits. */
    val memberId: String? get() = userId ?: ghostId
    val isGhost: Boolean get() = role == "ghost"
}

data class TripInvite(
    val email: String = "",
    val invitedBy: String = "",
    val status: String = "pending", // "pending" | "accepted" | "declined"
    val ghostId: String? = null,
)

data class CustomCategory(
    val id: String = "",
    val name: String = "",
    val emoji: String = "",
    val createdBy: String = "",
    val createdAt: Timestamp? = null,
)

/**
 * trips/{tripId}. Adds [memberUids] — a top-level array of real member uids — so the
 * app can query `whereArrayContains("memberUids", uid)` instead of the old
 * fetch-all-and-filter approach. Kept in sync with [members] on every mutation.
 */
data class Trip(
    val tripId: String = "",
    val name: String = "",
    val destination: String = "",
    val destinationCurrency: String = "USD",
    val startDate: String = "",
    val endDate: String = "",
    val createdBy: String = "",
    val members: List<TripMember> = emptyList(),
    val memberUids: List<String> = emptyList(),
    val invites: List<TripInvite> = emptyList(),
    // Emails with a pending invite — mirrors memberUids so Security Rules can verify
    // that a joining user was actually invited (request.auth.token.email in inviteEmails).
    val inviteEmails: List<String> = emptyList(),
    val customCategories: List<CustomCategory> = emptyList(),
    val status: String = "active", // "active" | "settled" | "archived"
    val createdAt: Timestamp? = null,
    val updatedAt: Timestamp? = null,
) {
    val realMembers: List<TripMember> get() = members.filter { it.role != "ghost" }
    val isOwner: (String) -> Boolean get() = { uid -> createdBy == uid }
}
